plugins/bkpr/bookkeeper.c

```diff
@@ -25,6 +27,158 @@ static struct db *db ;
+	json_add_u64(out, "timestamp", ev->timestamp);
+	json_object_end(out);
+}
+
+static void json_add_chain_event(struct json_stream *out,
+				 struct chain_event *ev)
+{
+	json_object_start(out, NULL);
+	json_add_string(out, "account", ev->acct_name);
+	json_add_string(out, "type", "chain");
+	json_add_string(out, "tag", ev->tag);
+	json_add_amount_msat_only(out, "credit", ev->credit);
+	json_add_amount_msat_only(out, "debit", ev->debit);
+	json_add_string(out, "currency", ev->currency);
+	json_add_outpoint(out, "outpoint", &ev->outpoint);
+	if (ev->spending_txid)
+		json_add_txid(out, "txid", ev->spending_txid);
+	if (ev->payment_id)
+		json_add_sha256(out, "payment_id", ev->payment_id);
+	json_add_u64(out, "timestamp", ev->timestamp);
+	json_add_u32(out, "blockheight", ev->blockheight);
+	json_object_end(out);
+}
+
+static void json_add_onchain_fee(struct json_stream *out,
+				 struct onchain_fee *fee)
+{
+	json_object_start(out, NULL);
+	json_add_string(out, "account", fee->acct_name);
+	json_add_string(out, "type", "onchain_fee");
+	json_add_string(out, "tag", "onchain_fee");
+	json_add_amount_msat_only(out, "credit", fee->credit);
+	json_add_amount_msat_only(out, "debit", fee->debit);
+	json_add_string(out, "currency", fee->currency);
+	json_add_u64(out, "timestamp", fee->timestamp);
+	json_add_txid(out, "txid", &fee->txid);
+	json_object_end(out);
+}
+
+/* Find all the events for this account, ordered by timestamp */
+static struct command_result *json_list_account_events(struct command *cmd,
+						       const char *buf,
+						       const jsmntok_t *params)
+{
+	struct json_stream *res;
+	struct account *acct;
+	const char *acct_name;
+	struct channel_event **channel_events;
+	struct chain_event **chain_events;
+	struct onchain_fee **onchain_fees;
+
+	if (!param(cmd, buf, params,
+		   p_opt("account", param_string, &acct_name),
+		   NULL))
+		return command_param_failed();
+
+	if (acct_name) {
+		db_begin_transaction(db);
+		acct = find_account(cmd, db, acct_name);
+		db_commit_transaction(db);
+
+		if (!acct)
+			return command_fail(cmd, JSONRPC2_INVALID_PARAMS,
+					    "Account '%s' not found",
+					    acct_name);
+	} else
+		acct = NULL;
+
+	db_begin_transaction(db);
+	if (acct) {
+		channel_events = account_get_channel_events(cmd, db, acct);
+		chain_events = account_get_chain_events(cmd, db, acct);
+		onchain_fees = account_get_chain_fees(cmd, db, acct);
+	} else {
+		channel_events = list_channel_events(cmd, db);
+		chain_events = list_chain_events(cmd, db);
+		onchain_fees = list_chain_fees(cmd, db);
+	}
+	db_commit_transaction(db);
+
+	res = jsonrpc_stream_success(cmd);
+	json_array_start(res, "events");
+	for (size_t i = 0, j = 0, k = 0;
+	     i < tal_count(channel_events)
+	     || j < tal_count(chain_events)
+	     || k < tal_count(onchain_fees);
+	     /* Incrementing happens inside loop */) {
+		struct channel_event *chan;
+		struct chain_event *chain;
+		struct onchain_fee *fee;
+		u64 lowest = 0;
```

Comment:
Why not set this to UINT64_MAX? Allows you to treat the "no lowest" case identical to "higher lowest" case.

In addition is seems that this assumes that there'll always be a fee to add at the end, wouldn't it be easier to do a merge-sort like structure where each iteration pops the lowest and prints that before restarting the iteration? This code might get the ordering wrong if we have two consecutive events of the same type followed by a later event, since this will at most print one event for each type before switching to the next type (though there may be a structure in the model I'm not considering here).

Reply:
> this will at most print one event for each type before switching to the next type

We only print one item every loop, there are `continue`s at each print action that ensures that there's only one per.

> this assumes that there'll always be a fee to add at the end

Since we continue after print, you'll never reach the fee print line if you're out of fees -- another item will have printed and the loop continued.

Generally speaking, the routine here is:
- elect the top item of each list
- from all (still populated) elected items, which has the lowest timestamp?
- print that (with some preference for channel, then chain, then fees in case of a tie)
- return to election phase

I could merge the lowest timestamp identification and 'topmost item' selection.